### ISSUE TITLE
Support of container flavors

### ISSUE THREAD

**Issue by huang195** [kubernetes/kubernetes#26369]
We want to enforce that pods can be only be created from a known set of flavors. I've started the work to create an admission control plugin to do the enforcement. Initially, it will only be enforcing cpu+memory, which is the same set of resources that quota and limitrange plugins manage today. 

@derekwaynecarr do you have any specific suggestions?

**Comment by derekwaynecarr** [comment]
@huang195 - can you elaborate what you mean by flavors?

Are you implying you want to support things like:

type1: requests=cpu=100m,memory=2Gi
type2: requests=cpu=1,memory=4Gi
...

And then only support pods that match a specific resource requirement?

**Comment by huang195** [comment]
Yeah, exactly.

**Comment by huang195** [comment]
Additionally, when user specifies a "partial" request, e.g., request=cpu=100m, it will fill the complete flavor as requests=cpu=100m,memory=2Gi.

**Comment by huang195** [comment]
> And then only support pods that match a specific resource requirement?
Clarification: I meant container flavors only, not pods.

**Comment by huang195** [comment]
@derekwaynecarr friendly ping, any other suggestions?

**Comment by derekwaynecarr** [comment]
I understand the rationale behind the request.  For background, in our OpenShift v2 product we had a similar concept with what we called gear sizes, but with the move to Kubernetes in our v3 solution, we have not built a similar concept yet.

As for things to think about:

* does a cluster-admin define container-shapes globally or per namespace?
* how would this integrate if we ever add a vertical auto-scaler in the future?

On a related note, I have been thinking about what a `LimitRange` would be in the future, and the container flavor concept is a bit like scoping a LimitRange to a particular subset of pod or container definitions.  Personally, I would like to support improvements in letting me target defaulting to a particular subset of pods (similar to scopes on `ResourceQuota`, but would also like to use field selectors which the project has not yet aligned on),

If you do want to prototype something for your own needs, you can probably do something with a `ThirdPartyResource` type and a small admission plugin.

**Comment by mikedanese** [comment]
This seems like it could be very easily solved in a configuration management system that sits on top of k8s. I don't see a strong reason to include this.

**Comment by huang195** [comment]
@derekwaynecarr that's very interesting. I'm glad you mentioned that you guys had something similar in openshift to deal with different flavors. Can you elaborate a bit more on why this is gone in v3? Are people not using it, finding it too hard to use, or what? I'm curious since Kubernetes currently doesn't support it natively. As @mikedanese mentioned, it's expecting an external entity to manage it.

**Comment by huang195** [comment]
@mikedanese I'm not sure I fully understand your comment. Admission control plugins such as quota and limitrange can also be done in a configuration management system that sits on top of k8s. What's the difference between flavors plugin and the existing plugins?

**Comment by derekwaynecarr** [comment]
@huang195 - right now, I tend to agree with @mikedanese , but I see no reason why you can't prototype the idea using `ThirdPartyResource`.  In general, there a huge amount of variations in how users want to assign resources, flavors is just one choice on the menu, and its not clear what your ultimate goal is for pursuing this model.

* Is your primary goal to improve your user experience?
* Have you looked at the `InitialResources` admission controller work? 
 * https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/initialresources/admission.go#L50
* Are you just trying to control the shape of a pod?  
 * Are the min/max features in LimitRange and the burstable ratio insufficient for you?

In OpenShift, each deployment and need is a bit unique.

* We support environments where users want to set both request and limits and admins can control shape and burstable ratio using LimitRange.
* We support environments that allow a cluster operator to run with a desired overcommit ratio. In that environment, our UI only exposes limits in our UI and requests are set server side based on the desired overcommit factor (https://github.com/openshift/origin/tree/388478c40e751c4295dcb9a44dd69e5ac65d0e3b/pkg/quota/admission/clusterresourceoverride)
* Finally, we have templates that can come pre-populated with compute resource requests/limits that users provision in our UI.

So basically, there are a lot of ways that folks choose to support similar concepts.
